File: go/654.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func constructMaximumBinaryTree(nums []int) *TreeNode {
	return cons(nums, 0, len(nums))
}

func cons(nums []int, l int, r int) *TreeNode {
	if l == r {
		return nil
	}
	max := max(nums, l, r)
	root := &TreeNode{Val: nums[max]}
	root.Left = cons(nums, l, max)
	root.Right = cons(nums, max+1, r)
	return root
}

func max(nums []int, l int, r int) int {
	id := l
	for i := l; i < r; i++ {
		if nums[id] < nums[i] {
			id = i
		}
	}
	return id
}

File: java/654.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode constructMaximumBinaryTree(int[] nums) {
        int l = nums.length;
        return helper(nums,0,l-1);
    }
    
    public TreeNode helper(int[] nums,int l,int r){
        if(l > r){
            return null;
        }
        if(l == r){
            return new TreeNode(nums[l]);
        }
        int max = l;
        for(int i = l;i <= r;i++){
            if(nums[i] > nums[max]){
                max = i;
            }
        }
        TreeNode node = new TreeNode(nums[max]);
        node.left = helper(nums,l,max-1);
        node.right = helper(nums,max+1,r);
        return node;
    }
}